accounting/src/delta/delta_data_collection/mod.rs

```diff
@@ -14,52 +14,94 @@
 // limitations under the License.
 
 pub mod undo;
+use self::undo::*;
 
 use std::collections::BTreeMap;
 
 use serialization::{Decode, Encode};
+use utils::ensure;
 
 use crate::error::Error;
 
-use self::undo::*;
-
-/// The outcome of combining two deltas for a given key upon the map that contains it
-#[derive(PartialEq, Eq, Debug)]
+/// The outcome of combining two deltas for a given key upon the map that contains it.
 enum DeltaMapOp<T> {
-    /// Write a specific value (for example, to write a Create or Modify operation)
-    Write(T),
-    /// Erase the value at the relevant key spot (for example, a modify followed by Erase yields nothing)
-    Delete,
+    /// Insert a specific value (for example, to write a Create or Modify operation)
+    Insert(DeltaMapElement<T>),
+    /// Erase the value at the relevant key spot (for example, a Create followed by Delete yields nothing)
+    Erase,
 }
 
 #[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
 pub enum DataDelta<T> {
-    Create(Box<T>),
-    Modify(Box<T>),
-    Delete,
+    // Stores new value
+    Create(T),
+    // Stores prev and new values
+    Modify(T, T),
+    // Stores value to restore
+    Delete(T),
 }
 
 #[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
-pub struct DeltaDataCollection<K: Ord, T> {
-    data: BTreeMap<K, DataDelta<T>>,
+pub enum DeltaMapElement<T> {
+    Delta(DataDelta<T>),
+    DeltaUndo(DataDeltaUndo<T>),
+}
+
+impl<T> DeltaMapElement<T> {
+    pub fn get_data_delta(&self) -> &DataDelta<T> {
+        match self {
+            DeltaMapElement::Delta(d) => d,
+            DeltaMapElement::DeltaUndo(d) => &d.0,
+        }
+    }
 }
 
-impl<K: Ord + Copy, T> DeltaDataCollection<K, T> {
+pub enum GetDataResult<T> {
```

Comment:
I'm wondering whether this is necessary... can't we learn this from the map itself without having this intermediary type?